init: upgrade to v12 (#607)
inveniosoftware/docs-invenio-rdm: docs/releases/upgrading/upgrade-v12.0.md

The only way to solve it is to destroy the index and reinit and
rebuild it from scratch. Yes the statistics will be only in the stored only in
the index but the statistic indices are created by templates and therefore not
affected by `invenio index destroy --yes-i-know`. The statistic indices will
remain in the list of indices and are totally functional also after a reinit.

### Data migration

Execute the data migration, note that there is no need to re-index the data:

```bash
pipenv run invenio shell $(find $(pipenv --venv)/lib/*/site-packages/invenio_app_rdm -name migrate_11_0_to_12_0.py)
```

### New roles

```bash
invenio roles create administration-moderation
invenio roles create administration

invenio access allow administration-moderation role administration-moderation
invenio access allow administration-access role administration
invenio access allow superuser-access role administration
```

### New configuration variables

```bash
COMMUNITIES_IDENTITIES_CACHE_REDIS_URL = "URI_TO_REDIS"
USERS_RESOURCES_ADMINISTRATION_ENABLED = True

> ```

Reviewer: see 

* https://github.com/inveniosoftware/cookiecutter-invenio-rdm/pull/273 
* https://github.com/inveniosoftware/invenio-theme/pull/382
:)

# Upgrading from v11 to v12

## Prerequisites

The steps listed in this article require an existing local installation of InvenioRDM v11.

!!! warning "Backup"

    Always backup your database and files before you try to perform an upgrade.

!!! info "Older Versions"

    In case you have an InvenioRDM installation older than v11, you can gradually upgrade

>     using the existing infrastructure to v11 and afterwards continue from here.

Reviewer 2: ```suggestion
    using the existing infrastructure to v11 and afterward continue from here.
```

Reviewer: depends on whether we're generally using American (afterward) or British (afterwards) English

    ```bash
    $ invenio-cli --version
    invenio-cli, version 1.
    ```

!!! info "Virtual environments"

    In case you are not inside a virtual environment, make sure that you prefix each `invenio`
    command with `pipenv run`.

**Local development**

Changing the Python version in your development environment highly
depends on your setup, and there is no golden rule.
One example would be to use [PyEnv](https://github.com/pyenv/pyenv).

You should delete your virtualenv before running `invenio-cli` or `pipenv` commands below.

!!! warning "Risk of losing data"

    Your virtual env folder contains uploaded files in InvenioRDM, in `var/instance/data`.
    If you need to keep such files, make sure you copy them over to the new virtual env in the same location.

### Upgrade InvenioRDM

(TODO: should we write that 3.9<=X<=3.12 is possible?)
Make sure that your virtual env is now running with Python 3.9.

Upgrade the RDM version:

```bash
cd <my-site>
# Upgrade to InvenioRDM v12
invenio-cli packages update 12.0.0
pipenv uninstall flask-babelex
# Re-build assets
invenio-cli assets build
```

> Optionally, update the file `<my-site>/Pipfile`. Attention: this action might lead to

Reviewer 3: I think if we remove the pre-release true restriction, we can also remove this ling

installing unwanted pre-releases of other packages.

```diff
[packages]
---invenio-app-rdm = {extras = [...], version = "~=11.0.0"}
+++invenio-app-rdm = {extras = [...], version = "~=12.0.0"}

> 

Reviewer 2: This change was a breaking change in pipenv package. Should we include it in the upgrade recipe?

```diff
- my-site = {editable="True", path="./site"}
+ my-site = {editable=true, path="./site"}
```

Reviewer 3: Yeah, it would be good to add, even though it's not technically a v12 change

[pipenv]

> allow_prereleases = true

Reviewer 3: I don't think this is needed; v12.0.0rc2 seems to install fine without it.

```

### Database migration

Execute the database migration:

```bash
# Execute the database migration
invenio alembic upgrade
```

### Declare usage statistics processing queues

```shell
invenio queues declare
```

### Update indices mappings

```shell
invenio index update communities-communities-v1.0.0
invenio index update rdmrecords-drafts-draft-v5.0.0
invenio index update rdmrecords-records-record-v5.0.0
```

> CAVEAT: this is not working because of an permission problem in

Reviewer 2: ```suggestion
CAVEAT: this is not working because of a permission problem in
```

[invenio-search](https://github.com/inveniosoftware/invenio-search/blob/d8b23ecf48f63d8d313f90fd4618a480e15fbd7b/invenio_search/ext.py#L448).
The only way to solve it is to destroy the index and reinit and

> rebuild it from scratch. Yes the statistics will be only in the stored only in
> the index but the statistic indices are created by templates and therefore not

Reviewer 2: ```suggestion
rebuild it from scratch. Yes the statistics will be only stored in
the index but the statistic indices are created by templates and therefore not
```

affected by `invenio index destroy --yes-i-know`. The statistic indices will
remain in the list of indices and are totally functional also after a reinit.

### Data migration

Execute the data migration, note that there is no need to re-index the data:

```bash
pipenv run invenio shell $(find $(pipenv --venv)/lib/*/site-packages/invenio_app_rdm -name migrate_11_0_to_12_0.py)
```

### New roles

```bash
invenio roles create administration-moderation
invenio roles create administration

invenio access allow administration-moderation role administration-moderation
invenio access allow administration-access role administration
invenio access allow superuser-access role administration
```

### New configuration variables

```bash

> COMMUNITIES_IDENTITIES_CACHE_REDIS_URL = "URI_TO_REDIS"

Reviewer 2: Should we make it more clear here: 
`COMMUNITIES_IDENTITIES_CACHE_REDIS_URL ="redis://cache:6379/4` 
and say something like: Replace `cache` with your own URI, so that end users can directly copy and paste the configuration, making it operational without the need to search for the correct value.

USERS_RESOURCES_ADMINISTRATION_ENABLED = True
THEME_SITENAME = "Project name for header and UI"
```

## Big Changes

- remove: dependency of flask-babelex
- add: concept doi (aka parent doi)
- add: statistics. (NOTE: statistic is stored in opensearch/libresearch)
- add: administration panel
- add: set quota
- add: branded communities

> 

Reviewer 3: licenses: The delimiter for the licenses vocabulary file has been changed from ";" to ",". If you use a custom vocabulary file you will need to update the delimiter if you want to update licenses in the future. The default licenses are available [here](https://github.com/inveniosoftware/invenio-rdm-records/blob/master/invenio_rdm_records/fixtures/data/vocabularies/licenses.csv). You can now add and update licenses with `pipenv run invenio rdm-records add-to-fixture licenses`

# Upgrading from v11 to v12

## Prerequisites

The steps listed in this article require an existing local installation of InvenioRDM v11.

!!! warning "Backup"

    Always backup your database and files before you try to perform an upgrade.

!!! info "Older Versions"

    In case you have an InvenioRDM installation older than v11, you can gradually upgrade
    using the existing infrastructure to v11 and afterwards continue from here.

## Upgrade Steps

!!! warning "Upgrade your invenio-cli"

    Make sure you have the latest `invenio-cli` installed. For InvenioRDM v12 it
    is at least v1.3.1

    ```bash
    $ invenio-cli --version
    invenio-cli, version 1.3.1
    ```

!!! info "Virtual environments"

    In case you are not inside a virtual environment, make sure that you prefix each `invenio`
    command with `pipenv run`.

**Local development**

Changing the Python version in your development environment highly
depends on your setup, and there is no golden rule.
One example would be to use [PyEnv](https://github.com/pyenv/pyenv).

> You should delete your virtualenv before running `invenio-cli` or `pipenv` commands below.

Reviewer 3: I would start this with, "If you upgraded your python version, "

!!! warning "Risk of losing data"

    Your virtual env folder contains uploaded files in InvenioRDM, in `var/instance/data`.
    If you need to keep such files, make sure you copy them over to the new virtual env in the same location.

### Upgrade InvenioRDM

Make sure that your virtual env is now running with at least Python 3.9. (InvenioRDM v12
is also tested with Python 3.12.)

There are two options to upgrade your system.

#### Option 1

This approach upgrades the dependencies in place. Your virtual environment for the
v11 version will be gone afterwards.

```bash
cd <my-site>

# Upgrade to InvenioRDM v12
invenio-cli packages update 12.0.0
pipenv uninstall flask-babelex

# Re-build assets
invenio-cli assets build
```

> #### Option 2

Reviewer 4: ```suggestion
#### Upgrade option 2: New virtual environment
```

option you chose.

##### Step 1
- create a new virtual environment
- activate your new virtual environment
- install `invenio-cli` by `pip install invenio-cli`

##### Step 2
Update the file `<my-site>/Pipfile`.

```diff
[packages]
---invenio-app-rdm = {extras = [...], version = "~=11.0.0"}
+++invenio-app-rdm = {extras = [...], version = "~=12.0.0"}
```

Due to a dependency upgrade, update the following line.
```diff
---my-site = {editable="True", path="./site"}
+++my-site = {editable=true, path="./site"}
```

##### Step 3
Update the `Pipfile.lock` file

```bash
invenio-cli packages lock
```

##### Step 4
install InvenioRDM v12

```bash
invenio-cli install
```

### Database migration

Execute the database migration:

```bash
# Execute the database migration
invenio alembic upgrade
```

### Declare usage statistics processing queues

```shell
invenio queues declare
```

### Data migration

Execute the data migration

```bash
pipenv run invenio shell $(find $(pipenv --venv)/lib/*/site-packages/invenio_app_rdm -name migrate_11_0_to_12_0.py)
```

### Reindex

```bash
invenio index destroy --yes-i-know
invenio index init
invenio rdm rebuild-all-indices
```

> In v12 the statistics will be stored in the index. The statistic indices are
> created by templates and therefore not affected by `invenio index destroy
> --yes-i-know`. The statistic indices will remain in the list of indices and are
> totally functional also after a reinit.

Reviewer 4: ```suggestion
In v12, record statistics will be stored in search indices rather than the database.
These indices are created based on *index templates* which are not affected by affected by `invenio index destroy --yes-i-know`.
So while this command will wipe the data and reset the numbers, the feature will stay functional after a re-init.
```
A bit of clarification from my current understanding, but I haven't tested this right now...

Author: i think your suggestion could cause some understanding problems `so while this command will wipe the data and reset the number` could be interpreted in a way that also the statistic indices will be removed. this is not the case. the statistic indices will stay also if i do a `invenio index destroy --yes-i-know`. tried in test and production.

Reviewer 4: Ah, I see, thanks for the clarification!
I've updated the suggestion, is it more accurate now?

### New roles

```bash
invenio roles create administration-moderation
invenio roles create administration

invenio access allow administration-moderation role administration-moderation
invenio access allow administration-access role administration
invenio access allow superuser-access role administration
```

### New configuration variables

```bash
COMMUNITIES_IDENTITIES_CACHE_REDIS_URL = "URI_TO_REDIS"
USERS_RESOURCES_ADMINISTRATION_ENABLED = True
THEME_SITENAME = "Project name for header and UI"
```

## Big Changes

- remove: dependency of flask-babelex
- add: concept doi (aka parent doi)
- add: statistics. (NOTE: statistic is stored in opensearch/libresearch)
- add: administration panel
- add: set quota
- add: branded communities

> - add: share options for drafts and records

Reviewer 3: add: built-in optional metadata fields

# Upgrading from v11 to v12

## Prerequisites

The steps listed in this article require an existing local installation of InvenioRDM v11.

!!! warning "Backup"

    Always backup your database and files before you try to perform an upgrade.

!!! info "Older Versions"

    In case you have an InvenioRDM installation older than v11, you can gradually upgrade
    using the existing infrastructure to v11 and afterwards continue from here.

## Upgrade Steps

!!! warning "Upgrade your invenio-cli"

    Make sure you have the latest `invenio-cli` installed. For InvenioRDM v12 it

>     is at least v1.2.0

Reviewer 3: Any reason not to tell folks to go to 1.3.1?

Author: i think at the point i wrote that, v1.2.0 was the newest. i will change that!

    ```bash
    $ invenio-cli --version
    invenio-cli, version 1.2.0
    ```

!!! info "Virtual environments"

    In case you are not inside a virtual environment, make sure that you prefix each `invenio`
    command with `pipenv run`.

**Local development**

> Changing the Python version in your development environment highly

Reviewer 3: I think this is leftover. If folks are on v11, they should already have python 3.9 and won't need to change.

Author: if they are then they can skip that step, if they are not than this is a really necessary step. i would keep it.

Reviewer 3: I would then start it with something like. "Python 3.9 or Python 3.12 are required to run InvenioRDM."

depends on your setup, and there is no golden rule.
One example would be to use [PyEnv](https://github.com/pyenv/pyenv).

You should delete your virtualenv before running `invenio-cli` or `pipenv` commands below.

!!! warning "Risk of losing data"

    Your virtual env folder contains uploaded files in InvenioRDM, in `var/instance/data`.
    If you need to keep such files, make sure you copy them over to the new virtual env in the same location.

### Upgrade InvenioRDM

Make sure that your virtual env is now running with at least Python 3.9. (InvenioRDM v12
is also tested with Python 3.12.)

There are two options to upgrade your system.

#### Option 1

This approach upgrades the dependencies in place. Your virtual environment for the
v11 version will be gone afterwards.

```bash
cd <my-site>

# Upgrade to InvenioRDM v12
invenio-cli packages update 12.0.0
pipenv uninstall flask-babelex

# Re-build assets
invenio-cli assets build
```

#### Option 2

> This approach will create a new virtual environment and leafs the v11 as it is.

Reviewer 4: ```suggestion
This approach will create a new virtual environment and leaves the v11 as it is.
```

If you are using a docker image on your production instance this will be the
option you chose.

##### Step 1
- create a new virtual environment
- activate your new virtual environment
- install `invenio-cli` by `pip install invenio-cli`

##### Step 2
Update the file `<my-site>/Pipfile`.

```diff
[packages]
---invenio-app-rdm = {extras = [...], version = "~=11.0.0"}
+++invenio-app-rdm = {extras = [...], version = "~=12.0.0"}
```

Due to a dependency upgrade, update the following line.
```diff
---my-site = {editable="True", path="./site"}
+++my-site = {editable=true, path="./site"}
```

##### Step 3
Update the `Pipfile.lock` file

```bash
invenio-cli packages lock
```

##### Step 4
install InvenioRDM v12

```bash
invenio-cli install
```

### Database migration

Execute the database migration:

```bash
# Execute the database migration
invenio alembic upgrade
```

### Declare usage statistics processing queues

```shell
invenio queues declare
```

### Update indices mappings

```shell
invenio index update communities-communities-v1.0.0
invenio index update rdmrecords-drafts-draft-v5.0.0
invenio index update rdmrecords-records-record-v5.0.0
```

> CAVEAT: this is not working because of a permission problem in

Reviewer 3: So should we just provide instructions to destroy the index? I don't see why we'd provide commands that don't work.

Author: the instructions are some lines below the question is if we should remove that part completely. i will remove the whole `update indices` section including the text. it is not necessary for the migration to explain that the `index index destroy` keeps the statistics because at this point no statistics exists but i will move that to the section `reindex` so that the information is there. in my opinion it is a important information which could be overlooked if it is written somewhere else and here it will be seen because the upgrade steps have to be gone through step by step.

destroy --yes-i-know`. The statistic indices will remain in the list of indices
and are totally functional also after a reinit.

### Data migration

Execute the data migration

```bash
pipenv run invenio shell $(find $(pipenv --venv)/lib/*/site-packages/invenio_app_rdm -name migrate_11_0_to_12_0.py)
```

### Reindex

```bash
invenio index destroy --yes-i-know
invenio index init
invenio rdm rebuild-all-indices
```

### New roles

```bash
invenio roles create administration-moderation
invenio roles create administration

invenio access allow administration-moderation role administration-moderation
invenio access allow administration-access role administration
invenio access allow superuser-access role administration
```

### New configuration variables

```bash
COMMUNITIES_IDENTITIES_CACHE_REDIS_URL = "URI_TO_REDIS"
USERS_RESOURCES_ADMINISTRATION_ENABLED = True
THEME_SITENAME = "Project name for header and UI"
```

## Big Changes

- remove: dependency of flask-babelex
- add: concept doi (aka parent doi)
- add: statistics. (NOTE: statistic is stored in opensearch/libresearch)
- add: administration panel
- add: set quota

> - add: branded communities

Reviewer 3: add: share options for drafts and records

Author: done
